src/infra/core/src/repos/dataset_repository_s3.rs

```diff
@@ -220,45 +272,71 @@ impl DatasetRepository for DatasetRepositoryS3 {
         //   collision
         let head = match dataset
             .as_metadata_chain()
-            .append(seed_block.into(), AppendOpts::default())
+            .append(
+                seed_block.into(),
+                AppendOpts {
+                    // We are using head ref CAS to detect previous existence of a dataset
+                    // as atomically as possible
+                    check_ref_is: Some(None),
+                    ..AppendOpts::default()
+                },
+            )
             .await
         {
-            Ok(hash) => Ok(hash),
+            Ok(head) => head,
             Err(AppendError::RefCASFailed(_)) => {
-                // We are using head ref CAS to detect previous existence of a dataset
-                // as atomically as possible
-                Err(CreateDatasetError::NameCollision(NameCollisionError {
-                    alias: dataset_alias.clone(),
-                }))
+                unreachable!("The duplication protection should have triggered by now");
```

Comment:
CAS error is still possible if same user is pushing same dataset in two concurrent processes at once - they would race to write the `head` ref file. We shouldn't be crashing the server in such situation.

But I guess it wouldn't be a "NameCollision" in this case, just an optimistic concurrency error...

Reply:
Resolved as internal error